File: app/services/completed_loan_services/tab_property.rb
module CompletedLoanServices
  class TabProperty
    attr_accessor :loan, :subject_property

    def initialize(loan, subject_property)
      @loan = loan
      @subject_property = subject_property
    end

    def call
      return false unless subject_property
      return false unless subject_property_completed?
      return false unless purpose_completed?

      true
    end

    def subject_property_completed?
      return false unless subject_property.address.present?
      return false unless address_completed?
      return false unless subject_property.usage.present?
      return false if !subject_property.primary_residence? && subject_property.gross_rental_income.nil?

      true
    end

    def purpose_completed?
      return false unless loan.purpose.present?
      return false if loan.purchase? && !subject_property.purchase_price.present?
      return false if loan.refinance? && !refinance_completed?

      true
    end

    def refinance_completed?
      return false unless subject_property.original_purchase_price.present?
      return false unless subject_property.original_purchase_year.present?

      true
    end

    def address_completed?
      return false unless subject_property.address

      address = subject_property.address
      return false if address.street_address.blank? && address.city.blank? && address.state.blank? && address.street_address2.blank? && address.zip.blank?

      address.full_text.present?
    end
  end
end

File: client/assets/javascripts/mixins/CompletedLoanMixins/TabProperty.js
var TabProperty = {
  propertyCompleted: function(loan) {
    if(loan.subject_property === undefined) { return false; }
    if(!this.subjectPropertyCompleted(loan.subject_property)) { return false; }
    if(!this.purposeCompleted(loan)) { return false; }

    return true;
  },

  subjectPropertyCompleted: function(property) {
    if(property.usage === null) { return false; }
    if(property.usage !== "primary_residence" && property.gross_rental_income === null) { return false; }
    if(!this.addressCompleted(property.address)) { return false; }

    return true;
  },

  addressCompleted: function(address) {
    if(address == null || address == undefined) { return false; }
    if(address.street_address == null && address.city == null && address.state == null && address.street_address2 == null)
      return false;

    return true;
  },

  purposeCompleted: function(loan) {
    if(loan.purpose == null) { return false; }
    if(loan.purpose == "purchase" && loan.subject_property.purchase_price == null) {
      return false;
    }
    if(loan.purpose == "refinance" && this.loanDoesNotHaveRefinanceData(loan)) {
      return false;
    }

    return true;
  },

  loanDoesNotHaveRefinanceData: function(loan) {
    if(loan.subject_property.original_purchase_price == null || loan.subject_property.original_purchase_year == null)
      return true;

    return false;
  }
}
module.exports = TabProperty;